### PR TITLE
Extract reindexing logic from transport action

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This commit extracts the reindexing logic from the transport action so
that it can be incorporated into the persistent reindex work without
requiring the usage of the client.

### CODE REVIEW

**Comment by tbrooks8** [comment]
@ywelsch This is the initial PR for work to be merged into master. It is ready for review.

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkByScrollAction.java [hunk 1]
```diff
@@ -116,10 +118,18 @@
 
     public AbstractAsyncBulkByScrollAction(BulkByScrollTask task, boolean needsSourceDocumentVersions,
```
Comment (ywelsch) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkByScrollAction.java]: I would prefer to remove this constructor, and  just keep the one with the  two extra parameters. Makes it a more conscious choice for callers  to pass `null` as parameter

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/ReindexValidator.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+    static void checkRemoteWhitelist(CharacterRunAutomaton whitelist, RemoteInfo remoteInfo) {
+        if (remoteInfo == null) {
+            return;
+        }
+        String check = remoteInfo.getHost() + ':' + remoteInfo.getPort();
+        if (whitelist.run(check)) {
+            return;
+        }
+        String whiteList = TransportReindexAction.REMOTE_CLUSTER_WHITELIST.getKey();
```
Comment (ywelsch) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/ReindexValidator.java]: nit: rename this to `whiteListKey`, as it's not the actual whitelist

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/Reindexer.java [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+                long l = fromNumber.longValue();
+                // Check that we didn't round when we fetched the value.
+                if (fromNumber.doubleValue() != l) {
+                    throw new IllegalArgumentException(name + " may only be set to an int or a long but was [" + from + "]");
+                }
+                return l;
+            }
+        }
+
+        private static Map<String, Object> buildConfig(ScriptService scriptService, ReindexSslConfig sslConfig) {
```
Comment (ywelsch) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/Reindexer.java]: this method isn't used. Where did it come from?

**Review by ywelsch — COMMENTED** [review]
looking good. I've left one question and 2 nits.

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/Reindexer.java (hunk 3 above)
Comment (tbrooks8) [reply modules/reindex/src/main/java/org/elasticsearch/index/reindex/Reindexer.java]: Leftover. Removing.

**Review by ywelsch — APPROVED** [review]
LGTM

**Pull request merged**